llama.cpp

```diff
@@ -1828,19 +1826,149 @@ struct llama_model_loader {
         }
     }
 
-    std::string get_arch_name() const {
-        const auto kv = LLM_KV(LLM_ARCH_UNKNOWN);
+    private:
+    template <typename T> struct gk_get_arrlen  { T & output; };
+    template <typename TI, typename TO> struct gk_set_literal { TI & input; TO & output; };
+    template <typename T>
+    void gk_set(int kid, T & result) {
+        (void)result;
+        throw std::runtime_error(format("request for key id %d with unhandled result type: %s", kid, typeid(T).name()));
+    }
+
+    void gk_set(int k, uint8_t     & r) { r = gguf_get_val_u8  (ctx_gguf, k); }
+    void gk_set(int k, uint16_t    & r) { r = gguf_get_val_u16 (ctx_gguf, k); }
+    void gk_set(int k, uint32_t    & r) { r = gguf_get_val_u32 (ctx_gguf, k); }
+    void gk_set(int k, uint64_t    & r) { r = gguf_get_val_u64 (ctx_gguf, k); }
+    void gk_set(int k, int8_t      & r) { r = gguf_get_val_i8  (ctx_gguf, k); }
+    void gk_set(int k, int16_t     & r) { r = gguf_get_val_i16 (ctx_gguf, k); }
+    void gk_set(int k, int32_t     & r) { r = gguf_get_val_i32 (ctx_gguf, k); }
+    void gk_set(int k, int64_t     & r) { r = gguf_get_val_i64 (ctx_gguf, k); }
+    void gk_set(int k, float       & r) { r = gguf_get_val_f32 (ctx_gguf, k); }
+    void gk_set(int k, double      & r) { r = gguf_get_val_f64 (ctx_gguf, k); }
+    void gk_set(int k, bool        & r) { r = gguf_get_val_bool(ctx_gguf, k); }
+    void gk_set(int k, std::string & r) { r = std::string(gguf_get_val_str(ctx_gguf, k)); }
+
+    template<typename T>
+    typename std::enable_if<std::is_integral<T>::value, void>::type
+    gk_set(int k, struct gk_get_arrlen<T> & r) { r.output = gguf_get_arr_n(ctx_gguf, k); }
+
+    template<typename TI, typename TO>
+    void gk_set_lit(TI i, TO o) {
+        (void)i; (void)o;
+        throw std::runtime_error(format("gk_set_lit can't handle types: in=%s, out=%s",
+            typeid(TI).name(), typeid(TO).name()));
+    }
+
+    template<typename T>
+    typename std::enable_if<std::is_integral<T>::value, void>::type
+    gk_set_lit(const int64_t & i, T & o) { o = T(i); }
+
+    template<typename T>
+    typename std::enable_if<std::is_floating_point<T>::value, void>::type
+    gk_set_lit(const double & i, T & o) { o = T(i); }
+
+    template<typename T>
+    void gk_set_lit(const T & i, T & o) { o = i; }
+
+    public:
+    template<typename T>
+    bool get_key(const std::string & key, T & result, const bool required = false) {
+        const auto & tt = typeid(T);
+        enum gguf_type gt = GGUF_TYPE_COUNT;
+        enum llama_model_kv_override_type ot = LLAMA_KV_OVERRIDE_INT;
+        bool is_signed = false, can_override = true;
+        if (tt == typeid(uint8_t)) {
```

Comment:
Why do this at runtime when you know the type at compile time? C++'s type_traits is a header that provides a lot of structs with value specializations, e.g. std::is_same. You can do the same thing but specialize the value of e.g. gguf_type_traits<uint8_t>::gguf_type (and use std::is_signed<T>::value to check whether the type is signed).